File: scripts/scrape_kki.py
import asyncio
import json
from playwright.async_api import async_playwright

# KKÍ Team URLs
MEN_TEAM_URL = "https://kki.is/motamal/leikir-og-urslit/motayfirlit/Eitt-lid?league_id=191&season_id=130402&team_id=4760732#mbt:6-200$t&0=5"
WOMEN_TEAM_URL = "https://kki.is/motamal/leikir-og-urslit/motayfirlit/Eitt-lid?league_id=189&season_id=130422&team_id=4760162"

async def scrape_team_stats(page, url):
    print(f"Navigating to {url}...")
    await page.goto(url)
    
    # Wait for the Genius Sports widget to load
    # It usually contains a table with class 'mbt-table'
    try:
        await page.wait_for_selector(".mbt-table", timeout=15000)
    except:
        print("Timeout waiting for stats table. Genius widget might be slow.")
        return []

    # Extract player stats
    players = await page.evaluate('''() => {
        const rows = Array.from(document.querySelectorAll('.mbt-table tbody tr'));
        return rows.map(row => {
            const cols = row.querySelectorAll('td');
            if (cols.length < 5) return null;
            
            return {
                name: cols[0].innerText.trim(),
                number: cols[1].innerText.trim(),
                mpg: parseFloat(cols[4].innerText) || 0,
                ppg: parseFloat(cols[5].innerText) || 0,
                rpg: parseFloat(cols[6].innerText) || 0,
                apg: parseFloat(cols[7].innerText) || 0,
                eff: parseFloat(cols[cols.length - 1].innerText) || 0
            };
        }).filter(p => p !== null);
    }''')
    
    return players

async def main():
    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=True)
        page = await browser.new_page()
        
        print("Scraping Men's Team...")
        men_players = await scrape_team_stats(page, MEN_TEAM_URL)
        
        print("Scraping Women's Team...")
        women_players = await scrape_team_stats(page, WOMEN_TEAM_URL)
        
        import os
        data_file_path = os.path.join(os.getcwd(), 'src', 'data', 'haukar_league_data.json')
        
        try:
            with open(data_file_path, "r", encoding="utf-8") as f:
                league_data = json.load(f)
        except Exception as e:
            print(f"Error reading {data_file_path}: {e}")
            league_data = {}

        if men_players and "korfubolti_karla" in league_data:
            league_data["korfubolti_karla"]["player_stats"] = men_players
            
        if women_players and "korfubolti_kvenna" in league_data:
            league_data["korfubolti_kvenna"]["player_stats"] = women_players
        
        with open(data_file_path, "w", encoding="utf-8") as f:
            json.dump(league_data, f, ensure_ascii=False, indent=2)
            
        print("Successfully merged KKÍ stats into haukar_league_data.json")
        await browser.close()

if __name__ == "__main__":
    asyncio.run(main())

File: scripts/scrape_kki.js
import puppeteer from 'puppeteer';
import fs from 'fs';
import path from 'path';

const MEN_TEAM_URL = "https://kki.is/motamal/leikir-og-urslit/motayfirlit/Eitt-lid?league_id=191&season_id=130402&team_id=4760732";
const WOMEN_TEAM_URL = "https://kki.is/motamal/leikir-og-urslit/motayfirlit/Eitt-lid?league_id=189&season_id=130422&team_id=4760162";

const DATA_FILE_PATH = path.join(process.cwd(), 'src', 'data', 'haukar_league_data.json');

const args = process.argv.slice(2);
const isRecentOnly = args.includes('--recent-only');

function needsRecentUpdate(divisionMatches) {
  if (!divisionMatches || divisionMatches.length === 0) return true;
  const now = new Date();
  const today = new Date(now.getFullYear(), now.getMonth(), now.getDate()).getTime();
  
  for (const match of divisionMatches) {
    if (!match.date) continue;
    const matchDate = new Date(match.date).getTime();
    const diffDays = (today - matchDate) / (1000 * 60 * 60 * 24);
    
    if (diffDays >= -1 && diffDays <= 1) {
      return true;
    }
  }
  return false;
}

async function scrapeKKI() {
  let leagueData = {};
  try {
    const rawData = fs.readFileSync(DATA_FILE_PATH, 'utf-8');
    leagueData = JSON.parse(rawData);
  } catch (err) {
    console.error("Could not read league data:", err);
    process.exit(1);
  }

  if (isRecentOnly) {
      console.log("Running in --recent-only targeted polling mode for KKÍ");
  }

  const browser = await puppeteer.launch({ 
      headless: "new",
      args: ['--no-sandbox', '--disable-setuid-sandbox']
  });

  async function scrapeTeam(urlBase, division) {
    if (isRecentOnly && leagueData[division] && !needsRecentUpdate(leagueData[division].matches)) {
        console.log(`No recent matches found for ${division}. Skipping targeted polling.`);
        return false;
    }

    const page = await browser.newPage();
    try {
      // 1. Fetch Schedule
      console.log(`Fetching matches for ${division}...`);
      await page.goto(`${urlBase}#mbt:6-200$t&0=1`, { waitUntil: 'domcontentloaded' });
      await page.waitForSelector('.mbt-table', { timeout: 15000 });
      
      const matches = await page.evaluate(() => {
         const rows = Array.from(document.querySelectorAll('.mbt-table tbody tr'));
         return rows.map(row => {
             const cols = row.querySelectorAll('td');
             if (cols.length < 4) return null;
             
             let scoreIdx = -1;
             for(let i=0; i<cols.length; i++) {
                 if(cols[i].innerText.includes('-')) {
                     scoreIdx = i;
                     break;
                 }
             }
             if (scoreIdx === -1) scoreIdx = 3; // Fallback
             
             const dateText = cols[0]?.innerText.trim() || '';
             const home = cols[scoreIdx - 1]?.innerText.trim() || '';
             let score = cols[scoreIdx]?.innerText.trim() || '';
             const away = cols[scoreIdx + 1]?.innerText.trim() || '';
             
             // Format date: DD.MM.YYYY -> YYYY-MM-DD
             let formattedDate = dateText;
             const parts = dateText.split('.');
             if (parts.length === 3) {
                 formattedDate = `${parts[2]}-${parts[1].padStart(2, '0')}-${parts[0].padStart(2, '0')}`;
             }
             
             if (score.includes('-')) {
                 if (!score.match(/\d/)) score = null; // Unplayed game usually has "-"
             } else {
                 score = null;
             }
             
             return {
                 date: formattedDate,
                 home,
                 away,
                 score,
                 competition: "Besta deild"
             };
         }).filter(m => m !== null && (m.home.includes("Haukar") || m.away.includes("Haukar")));
      });
      
      if (matches && matches.length > 0) {
          if (!leagueData[division]) leagueData[division] = {};
          leagueData[division].matches = matches;
      }

      // 2. Fetch Player Stats
      console.log(`Fetching player stats for ${division}...`);
      await page.goto(`${urlBase}#mbt:6-200$t&0=5`, { waitUntil: 'domcontentloaded' });
      await page.waitForSelector('.mbt-table', { timeout: 15000 });

      const players = await page.evaluate(() => {
          const rows = Array.from(document.querySelectorAll('.mbt-table tbody tr'));
          return rows.map(row => {
              const cols = row.querySelectorAll('td');
              if (cols.length < 5) return null;
              
              return {
                  name: cols[0].innerText.trim(),
                  number: cols[1].innerText.trim(),
                  mpg: parseFloat(cols[4].innerText) || 0,
                  ppg: parseFloat(cols[5].innerText) || 0,
                  rpg: parseFloat(cols[6].innerText) || 0,
                  apg: parseFloat(cols[7].innerText) || 0,
                  eff: parseFloat(cols[cols.length - 1].innerText) || 0
              };
          }).filter(p => p !== null && p.name !== "Totals");
      });
      
      if (players && players.length > 0) {
          if (!leagueData[division]) leagueData[division] = {};
          leagueData[division].player_stats = players;
      }
      
      return true;
    } catch (e) {
      console.error(`Error scraping ${division}:`, e.message);
      return false;
    } finally {
      await page.close();
    }
  }

  const updatedMen = await scrapeTeam(MEN_TEAM_URL, "korfubolti_karla");
  const updatedWomen = await scrapeTeam(WOMEN_TEAM_URL, "korfubolti_kvenna");

  if (updatedMen || updatedWomen) {
      fs.writeFileSync(DATA_FILE_PATH, JSON.stringify(leagueData, null, 2), 'utf-8');
      console.log("Successfully merged KKÍ stats and matches into haukar_league_data.json");
  } else {
      console.log("KKÍ check complete. No updates needed.");
  }

  await browser.close();
}

scrapeKKI();
